pyquil/operator_estimation.py

```diff
@@ -479,3 +497,164 @@ def measure_observables(qc: QuantumComputer, tomo_experiment: TomographyExperime
+
+    Note that for the given example of X(1) * Z(0) and Z(1) * X(0), we can construct
+    the following basis which simultaneously diagonalizes both operators:
+
+      -- 2 * |0>' = |0> (|0> + |1>) + |1> (|0> - |1>)
+      -- 2 * |1>' = |0> (|0> + |1>) - |1> (|0> - |1>)
+      -- 2 * |2>' = |0> (|0> - |1>) + |1> (|0> + |1>)
+      -- 2 * |3>' = |0> (-|0> + |1>) + |1> (|0> + |1>)
+
+    In this basis, X Z looks like diag(1, -1, 1, -1), and Z X looks like diag(1, 1, -1, -1)
+
+    :param pauli_a: PauliTerm to check diagonality of in the natural tpb of `pauli_b`
+    :param pauli_b: PauliTerm to check diagonality of in the natural tpb of `pauli_a`
+    :return: Boolean of diagonality in each others natural tpb
+    """
+    overlapping_active_qubits = set(pauli_a.get_qubits()) & set(pauli_b.get_qubits())
+    for qubit_index in overlapping_active_qubits:
+        if (pauli_a[qubit_index] != 'I' and pauli_b[qubit_index] != 'I'
+                and pauli_a[qubit_index] != pauli_b[qubit_index]):
+            return False
+
+    return True
+
+
+def get_diagonalizing_basis(list_of_pauli_terms: List[PauliTerm]):
+    """
+    Find the Pauli Term with the most non-identity terms
+    :param list_of_pauli_terms: List of Pauli terms to check
+    :return: The highest weight Pauli Term
+    """
+    qubit_ops = set(reduce(lambda x, y: x + y,
+                           [list(term._ops.items()) for term in list_of_pauli_terms]))
+    qubit_ops = sorted(list(qubit_ops), key=lambda x: x[0])
+
+    return PauliTerm.from_list(list(map(lambda x: tuple(reversed(x)), qubit_ops)))
+
+
+def _max_key_overlap_term_pair(expt_setting: ExperimentSetting, diagonal_sets: Dict):
+    """
+    Calculate the max overlap of an ExperimentSetting with keys of diagonal_sets.
+    Returns a different key if we find any collisions. If no collisions are found,
+    then the ExperimentSetting is added (as a tuple) and the key is updated so it has the
+    largest weight.
+    :param expt_setting: ExperimentSetting
+    :param diagonal_sets: dictionary with
+        (key, value): ((diagonal_in_basis, diagonal_out_basis): list of tuples of PauliTerms
+                                                    that are diagonal in those bases)
+    :return: the updated diagonal_sets dictionary updated with the input expt_setting placed
+            appropriately into it
+    """
+    # construct a tuple out of the in and out operators of the ExperimentSetting
+    tup_pauli_terms = tuple(([expt_setting.in_operator], [expt_setting.out_operator]))
+    # loop through keys and determine if in_operator is diagonal in the same tpb as in_key,
+    # and out_operator the same with out_key
+    for key in list(diagonal_sets.keys()):
+        # obtain the in and out keys
+        in_key = key[0]
+        out_key = key[1]
+        # construct PauliTerms from in and out keys
+        pauli_in_from_key = PauliTerm.from_list([x[::-1] for x in in_key])
```

Comment:
this seems incredibly roundabout

Reply:
i.e. leave everything as PauliTerms

Reply:
Rewrote the entire function to work with `ExperimentSettings` more directly.